### PR TITLE
Grf qr engine

### PR DESCRIPTION
**Pull request by dajmcdon**
### Checklist

Please:

- [x] Make sure this PR is against "dev", not "main".
- [x] Request a review from one of the current epipredict main reviewers:
      dajmcdon.
- [x] Make sure to bump the version number in `DESCRIPTION` and `NEWS.md`.
      Always increment the patch version number (the third number), unless you are
      making a release PR from dev to main, in which case increment the minor
      version number (the second number).
- [x] Describe changes made in NEWS.md, making sure breaking changes
      (backwards-incompatible changes to the documented interface) are noted.
      Collect the changes under the next release number (e.g. if you are on
      0.7.2, then write your changes under the 0.8 heading).

### CODE REVIEW

**Review by dsweber2 — APPROVED** [review]
Comparing `make_quantile_reg` and `make_grf_quantiles`, it looks like they're both using the same model/engine/mode settings. I remember some discussion about maybe switching so that `quantile_regression` was a separate mode? I guess if that's happening it's happening some later version?

Also, I'm assuming all parameters not here are just set to their defaults, so `sample.fraction=0.5`; is there some reason for this particular trio of parameters, and not including say `sample.fraction`, `honesty`, `honesty.fraction`, `honesty.prune.leaves`, etc?

Overall, mostly just some clarification stuff and some bike-shedding w.r.t which parameters we include. Feel free to merge.

File: R/make_grf_quantiles.R [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+#'
+#' [grf::quantile_forest()] fits random forests in a way that makes it easy
+#' to calculate _quantile_ forests. Currently, this is the only engine
+#' provided here, since quantile regression is the typical use-case.
+#'
+#' @section Tuning Parameters:
+#'
+#' This model has 3 tuning parameters:
+#'
+#' - `mtry`: # Randomly Selected Predictors (type: integer, default: see below)
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply R/make_grf_quantiles.R:11]: This is just copied from the `parsnip::rand_forest()` docs.

File: R/make_grf_quantiles.R [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+#' - `trees`: # Trees (type: integer, default: 2000L)
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply R/make_grf_quantiles.R:12]: Same.

File: R/make_grf_quantiles.R [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+#' ewf <- epi_workflow(rec, spec, frost) %>%
+#'   fit(dat) %>%
+#'   forecast()
+#' ewf %>%
+#'   rename(forecast_date = time_value) %>%
+#'   mutate(target_date = forecast_date + 7L) %>%
+#'   pivot_quantiles_wider(.pred)
+#'
+#' @name grf_quantiles
+NULL
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply R/make_grf_quantiles.R:85]: `?grf_quantiles` should display this documentation (but nothing for `?make_grf_quantiles`).

File: R/make_quantile_reg.R [hunk 4]
```diff
@@ -106,15 +109,15 @@ make_quantile_reg <- function() {
     out <- switch(type,
       rq = dist_quantiles(unname(as.list(x)), object$quantile_levels), # one quantile
       rqs = {
-        x <- lapply(unname(split(x, seq(nrow(x)))), function(q) sort(q))
+        x <- lapply(vctrs::vec_chop(x), function(x) sort(drop(x)))
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply R/make_quantile_reg.R:112]: just more efficient. The purpose is to "split a matrix into a list by row". That's what the second does. The first is a very old-fashioned base-R implementation.

File: tests/testthat/test-grf_quantiles.R [hunk 5]
```diff
@@ -0,0 +1,52 @@
+set.seed(12345)
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply tests/testthat/test-grf_quantiles.R:1]: It should always work, but at least this way, I can track the values if necessary.

File: tests/testthat/test-grf_quantiles.R [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    quantiles = c(0.1, 0.5, 0.9), seed = fseed,
+    mtry = 2L, min.node.size = 10
+  )
+  p_pars <- predict(out, new_data = tib[1:5, ]) %>%
+    pivot_quantiles_wider(.pred)
+  p_manual <- predict(manual, newdata = as.matrix(tib[1:5, 2:3]))$predictions
+  colnames(p_manual) <- c("0.1", "0.5", "0.9")
+  p_manual <- tibble::as_tibble(p_manual)
+  # not equal despite the seed, etc
+  # expect_equal(p_pars, p_manual)
```
[the remark this answers is not in the document]
Comment (dajmcdon) [reply tests/testthat/test-grf_quantiles.R:31]: My understanding is that these two _should_ result in the same thing. So, I'm a bit miffed it doesn't.
``` r
library(epipredict)
library(grf)
tib <- tibble::tibble(
  y = rnorm(100), x = rnorm(100), z = rnorm(100),
  f = factor(sample(letters[1:3], 100, replace = TRUE))
)
fseed <- 12345
spec_seed <- rand_forest(mode = "regression", mtry = 2L, min_n = 10) %>%
  set_engine("grf_quantiles", seed = fseed)

manual <- quantile_forest(
  as.matrix(tib[, 2:3]), tib$y,
  quantiles = c(0.1, 0.5, 0.9), seed = fseed,
  mtry = 2L, min.node.size = 10
)
translate(spec_seed)
#> Random Forest Model Specification (regression)
#> 
#> Main Arguments:
#>   mtry = 2
#>   min_n = 10
#> 
#> Engine-Specific Arguments:
#>   seed = fseed
#> 
#> Computational engine: grf_quantiles 
#> 
#> Model fit template:
#> grf::quantile_forest(X = missing_arg(), Y = missing_arg(), mtry = min_cols(~2L, 
#>     x), min.node.size = min_rows(~10, x), seed = fseed, quantiles = c(0.1, 
#> 0.5, 0.9), num.threads = 1L)
```

<sup>Created on 2024-08-13 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

The function is being called with the same args, and passing the seed. I'm not concerned, but I wish the were the same.

**Comment by dajmcdon** [comment]
As for the overall comment, yes, the plan is for `{parsnip}` to add a `mode = "quantile_regression"` [setting](https://github.com/tidymodels/planning/pull/30#discussion_r1707273740). At that point, we'll probably pivot all of the engines here to that setup, and potentially remove them from here.

File: R/make_grf_quantiles.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (dsweber2) [reply R/make_grf_quantiles.R:85]: strange, it isn't after I `devtools::load_all` this package

[the remark this answers is not in the document]
Comment (dajmcdon) [reply R/make_grf_quantiles.R:85]: You may need to `devtools::document()`. It works for me.

[the remark this answers is not in the document]
Comment (dsweber2) [reply R/make_grf_quantiles.R:85]: that didn't do it, but at this point I'm going to chalk it up to state having to do with this being a PR, and call it good.

**Pull request merged**